### ISSUE TITLE
testing more external codebases before release.

### ISSUE THREAD

**Issue by Aladdin-ADD** [eslint/eslint#9334]
I'm wondering it is enough only linting ESLint codebase before release -- sometimes it cannot catch unexpected breaking changes. is it worthy to add more?

**Comment by platinumazure** [comment]
Hi @Aladdin-ADD, we created [eslint-canary](https://github.com/eslint/eslint-canary/) for more or less this purpose. Maybe we need to evaluate if we're using it optimally.

@eslint/eslint-team Thoughts? Are we using eslint-canary to the fullest? Would it have caught the eslint-plugin-react issue reported around the node parent change?

**Comment by ilyavolodin** [comment]
In the current configuration, it would not catch those errors, since it runs on a nightly bases, and haven't failed with the latest changes. My original regression build might've caught those changes, because it was running unit tests for `eslint-plugin-react`, but it was really tough to maintain it.

**Comment by Aladdin-ADD** [comment]
eh..I missed it. Some widely used plugins and parsers are worthy to be covered, I guess~

**Comment by jseminck** [comment]
Where is the info regarding the `eslint-plugin-react` discussion?

**Comment by kaicataldo** [comment]
I think we should be using `eslint-canary` for this purpose. What could we be doing better/differently?

@jseminck It doesn't look like an issue exists. If this is an issue for you would you like to make one? Looks like you might have more info/context around this :)

**Comment by Aladdin-ADD** [comment]
surely we should be using eslint-canary, I just have a few suggestions:

- more projects included,  especially widely used plugins such as `eslint-plugin-react`

- flexibility. these codebase is evolving, there would be some code reporting errors -- not our reasons, can be easily disabled.

- continually. these testing don't have to be in release. I think finding some errors earlier is better, and it won't put too much things on the release time.

- smarter/automatic tool. maintaining this project list would be tough.

**Comment by jseminck** [comment]
One thing that happened in ESLint is that a case for a rule that we previously didn't support, suddenly started working with ESLint 4.7.0. But this was expected.

Relevant PR for eslint-plugin-react where we ignored the error and moved the test-case to valid (even though it should be invalid): https://github.com/yannickcr/eslint-plugin-react/pull/1375

Issue which I created on ESLint: https://github.com/eslint/eslint/issues/9101

Current PR to fix our build (which was not yet accepted, because if we accept it the plugin will behave differently whether you have ESLint <= 4.6 or ESLint >= 4.7): https://github.com/yannickcr/eslint-plugin-react/pull/1437

All in all I wouldn't say this is a major issue, in fact the plugin became better as it now also detects a case that we ignored previously. That's why I am wondering if there were also any other issues?

This change could have only been detected by you (the ESLint team) if you had a way to build plugins (= run tests) with the new ESLint version before it is actually released though. But then again I am not sure that plugins can expect ESLint to verify there are no breaking changes for every release.

**Comment by kaicataldo** [comment]
@Aladdin-ADD Great, thanks for the input!

We currently run it nightly instead of on every CI build because it takes a while to run. Ideally anyone working on code would run it locally, but I'm not sure how we enforce/ask that (I know I don't always, if it's a smaller change).

Sounds like it might be worth having a follow up discussion around how we integrate `eslint-canary` into our workflow better! We started using a few months ago to see how it would go, and maybe it's time we re-evaluate and see what we think/how we can improve things.

@jseminck Thanks for all the info! Ah, understood - I didn't realize you actually made the original issue. We of course avoid breaking changes in semver-patch and semver-minor releases, however we do sometimes release changes that result in an increased number of warnings to fix buggy behavior (which I personally think this falls under). That being said, I don't think this was intended to be a breaking change - so I apologize about that.

I haven't seen any other issues related to this. cc @eslint/eslint-team

**Comment by not-an-aardvark** [comment]
@kaicataldo That change also caused https://github.com/eslint/eslint/issues/9331.

**Comment by ilyavolodin** [comment]
@jseminck As @not-an-aardvark mentioned, we reverted this due to #9331 we are still planning on adding this change back, but will wait until next major release, since this has potential to break some plugins (there were some that relied on `parent` property to not be on the nodes, I think).

As for `eslint-canary`, we had (a while back) a nightly build that would install most widely used plugins, pull latest ESLint code and run plugin's unit tests. Unfortunately, it started failing once some of the plugins started switching over to yarn and dropped support for npm, so we stopped maintaining it.
